ocular/src/keyring.rs

```diff
@@ -0,0 +1,752 @@
+use bip32::PrivateKey;
+use bip39::{Mnemonic, Seed};
+use ripemd::Ripemd160 as Ripemd;
+use serde::{Deserialize, Serialize};
+use sha2::Digest as Sha2Digest;
+use sha2::Sha256;
+use signatory::{
+    pkcs8::der::Document, pkcs8::EncodePrivateKey, pkcs8::LineEnding, pkcs8::PrivateKeyDocument,
+    FsKeyStore, KeyName,
+};
+use std::{env, fs, path::Path};
+
+use crate::error::KeyStoreError;
+
+// Constants
+// TODO: Move to independant constants file if reused elsewhere
+static COSMOS_BASE_DERIVATION_PATH: &str = "m/44'/118'/0'/0/0";
+static COSMOS_ADDRESS_PREFIX: &str = "cosmos";
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum KeyRingType {
+    File,
+    // TODO: OS (and other) Types
+}
+
+impl Default for KeyRingType {
+    fn default() -> Self {
+        KeyRingType::File
+    }
+}
+
+/// Basic keystore traits that all backends are expected to implement
+pub trait KeyStore {
+    /// Create new key store
+    fn create_key_store(&mut self) -> Result<(), KeyStoreError>;
+
+    /// Check if key store has been initialized.
+    fn key_store_created(&self) -> bool;
+
+    /// Check if key exists under specific name. Will return false if no key is found.
+    fn key_exists(&self, name: &str) -> Result<bool, KeyStoreError>;
+
+    /// Add a new key based off of name, password, and derivation path (defaults to cosmos). If override_if_exists is set to true, it will override any existing key with the same name.
+    fn add_key(
+        &self,
+        name: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<PrivateKeyOutput, KeyStoreError>;
+
+    /// Delete key with a given name. If no key exists under name specified an error will be thrown.
+    fn delete_key(&self, name: &str) -> Result<(), KeyStoreError>;
+
+    /// Rename key. If override_if_exists is true any keys with the new key name will be forecfully overriden.
+    fn rename_key(
+        &self,
+        current_name: &str,
+        new_name: &str,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError>;
+
+    /// Get key address in bech32 (aka segwit) format. Will throw an error if the key does not exist.
+    fn get_key_address(&self, name: &str) -> Result<PublicKeyOutput, KeyStoreError>;
+
+    /// Get all key addresses in bech32 (aka segwit) format.
+    fn get_all_keys(&self) -> Result<Vec<PublicKeyOutput>, Box<dyn std::error::Error>>;
+
+    /// Recover key via mnemonic, password, and derivation_path (defaults to cosmos). If override_if_exists is set to true, it will override any existing key with the same name.
+    fn recover_from_mnemonic(
+        &self,
+        name: &str,
+        mnemonic: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError>;
+}
+
+/// Mnemonic and private key in pkcs8 PrivateKeyDocument format
+#[derive(Debug)]
+pub struct PrivateKeyOutput {
+    pub mnemonic: String,
```

Comment:
You might want to use [`bip32::Mnemonic`](https://docs.rs/bip32/latest/bip32/struct.Mnemonic.html) here, which handles clearing the mnemonic out of memory on drop. Also it generally seems bad to persist the mnemonic for any longer than is strictly necessary to derive a key.

Reply:
Good point, updated to [bip32::Mnemonic](https://docs.rs/bip32/latest/bip32/struct.Mnemonic.html).